engines/director/lingo/lingo-bytecode.cpp

```diff
@@ -0,0 +1,311 @@
+					uint32 pointer = value;
+					while (pointer < constsStoreSize) {
+						if (constsStore[pointer] == '\r') {
+							constant.u.s += '\n';
+						} else if (constsStore[pointer] == '\0') {
+							break;
+						} else {
+							constant.u.s += constsStore[pointer];
+						}
+						pointer += 1;
+					}
+					if (pointer >= constsStoreSize) {
+						warning("Constant string has no null terminator");
+						break;
+					}
+				}
+				break;
+			case 4: // Integer type
+				constant.type = INT;
+				constant.u.i = (int)value;
+				break;
+			case 9:  // Float type
+				{
+					constant.type = FLOAT;
+					if (value < constsStoreOffset) {
+						warning("Constant float start offset is out of bounds");
+						break;
+					} else if (value + 4 > constsStoreSize) {
+						warning("Constant float end offset is out of bounds");
+						break;
+					}
+					constant.u.f = *(float *)(constsStore + value);
+				}
+				break;
+			default:
+				warning("Unknown constant type %d", type);
+				break;
+		}
+
+		constsData.push_back(constant);
+	}
+	free(constsStore);
+
+	// parse each function!
+	// these are stored as a code storage area, followed by a reference table of 42 byte entries.
+
+	// copy the storage area first.
+	uint32 codeStoreSize = functionsOffset - codeStoreOffset;
+	stream.seek(codeStoreOffset);
+	byte *codeStore = (byte *)malloc(codeStoreSize);
+	stream.read(codeStore, codeStoreSize);
+
+	// read each entry in the function table.
+	stream.seek(functionsOffset);
+	for (uint16 i=0; i<functionsCount; i++) {
+		_currentScriptFunction = i;
+		_currentScriptContext->functions.push_back(new ScriptData);
+		_currentScript = _currentScriptContext->functions[_currentScriptFunction];
+
+		uint16 nameIndex = stream.readUint16();
+		stream.readUint16();
+		uint32 length = stream.readUint32();
+		uint32 startOffset = stream.readUint32();
+		uint16 argCount = stream.readUint16();
+		uint32 argOffset = stream.readUint32();
+		uint16 varCount = stream.readUint16();
+		uint32 varNamesOffset = stream.readUint32();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+		stream.readUint16();
+
+		if (startOffset < codeStoreOffset) {
+			warning("Function %d start offset is out of bounds!", i);
+			continue;
+		} else if (startOffset + length >= codeStoreOffset + codeStoreSize) {
+			warning("Function %d end offset is out of bounds", i);
+			continue;
+		}
+
+		uint16 pointer = startOffset - codeStoreOffset;
+		Common::Array<uint32> offsetList;
+		while (pointer < startOffset + length - codeStoreOffset) {
+			uint8 opcode = codeStore[pointer];
+
+			pointer += 1;
+			if (_lingoV4.contains(opcode)) {
+				offsetList.push_back(_currentScript->size());
+				g_lingo->code1(_lingoV4[opcode]->func);
+				size_t argc = strlen(_lingoV4[opcode]->proto);
+				for (uint c=0; c < argc; c++) {
+					switch (_lingoV4[opcode]->proto[c]) {
```

Comment:
Another over-indented set of `case` statements.

Reply:
Fixed